src/QuadVelocityController.cpp

```diff
@@ -21,65 +21,81 @@
 roll_pid_(roll_pid[0], roll_pid[1], roll_pid[2], roll_pid[3], roll_pid[4]),
-yaw_pid_(yaw_pid[0], yaw_pid[1], yaw_pid[2], yaw_pid[3], yaw_pid[4]),
-last_transform_stamped_(),
-have_last_transform_(false),
+tfBuffer_(),
+tfListener_(tfBuffer_),
+odometry_subscriber_(nh.subscribe(
+                         "odometry/filtered",
+                         100,
+                         &QuadVelocityController::odometryCallback,
+                         this)),
+odometry_msg_queue_(),
+setpoint_(),
 hover_throttle_(hover_throttle),
-expected_update_frequency_(expected_update_frequency)
+max_initial_wait_(getParam<double>(private_nh, "max_initial_wait_seconds")),
+max_wait_(getParam<double>(private_nh, "max_wait_seconds"))
 {
-
 }
 
 // Take in a target velocity that does not take into account the quads current heading
 // And transform it to the velocity vectors that correspond to the quads current yaw
 // Set the PID's set points accordingly
 void QuadVelocityController::setTargetVelocity(geometry_msgs::Twist twist)
 {
-    // Throttle and yaw_pid need no correction.
-    // Remember these are velocities.
-    throttle_pid_.setSetpoint(twist.linear.z);
-    yaw_pid_.setSetpoint(twist.angular.z);
-
-    double last_yaw
-        = yawFromQuaternion(last_transform_stamped_.transform.rotation);
-
-    // Pitch and roll velocities are transformed according to the last_yaw
-    // angle because the incoming target velocities do not take the current yaw
-    // into account
-    pitch_pid_.setSetpoint(twist.linear.x * std::cos(last_yaw)
-                         + twist.linear.y * std::sin(last_yaw));
-
-    // Note: Roll is inverted because a positive y velocity means a negative
-    // roll by the right hand rule
-    roll_pid_.setSetpoint(-(twist.linear.x * -std::sin(last_yaw)
-                          + twist.linear.y * std::cos(last_yaw)));
+    setpoint_ = twist;
 }
 
 // Main update, runs all PID calculations and returns a desired uav_command
 // Needs to be called at regular intervals in order to keep catching the latest velocities.
 bool QuadVelocityController::update(const ros::Time& time,
                                     iarc7_msgs::OrientationThrottleStamped& uav_command)
 {
+    if (time < last_update_time_) {
+        ROS_ERROR("Tried to update QuadVelocityController with time before last update");
+        return false;
+    }
+
     // Get the current velocity of the quad.
-    geometry_msgs::Twist velocity;
-    bool success = getVelocityAtTime(velocity, time);
+    geometry_msgs::Vector3 velocity;
+    bool success = getVelocityAtTime(velocity, time, max_wait_);
     if (!success) {
         ROS_WARN("Failed to get current velocities in QuadVelocityController::update");
         return false;
     }
 
+    // Get the current transform (rotation) of the quad
+    geometry_msgs::TransformStamped transform;
+    success = getTransformAtTime(transform, time, max_wait_);
```

Comment:
Side note: we wouldn't need this and all that goes with it if we used robot localization to publish the orientation.